### PR TITLE
fix(penelope): add turn-aware prompts and simplify orchestration

### PR DESCRIPTION
**Pull request by harry-rhesis**
## Purpose

When `min_turns` is configured, Penelope achieves its test goal early but is forced to continue — and with no guidance on what to do with remaining turns, it resorts to filler ("thank you" messages, verbatim repetition, or evaluator-voice summaries like "This test has failed"). This PR makes the agent turn-budget-aware and fixes several evaluation bugs exposed along the way, then simplifies the resulting architecture.

## What changed

**Turn-budget awareness** (agent + prompts):
- System prompt gains a "Turn Budget Management" section with deepening strategies (verify consistency, probe edge cases, test adjacent scenarios, challenge previous answers, cross-reference)
- Subsequent turn prompt rendered with live turn context ("Turn 4 of 10, min 5")
- "Message Voice" section added to system prompt — DO/DON'T examples enforcing natural user voice, preventing evaluator-speak from leaking into messages
- Anti-repetition rules in both system prompt (source of truth) and per-turn prompt (reinforcement at moment of action)

**Evaluation fixes**:
- `GoalAchievedCondition`: decoupled "goal impossible" floor (80% of max_turns) from min_turns — agent can declare success at min_turns but can't give up until 80% through the budget
- Agent passes original test instructions to goal judge, not rendered instructions containing turn-budget directives — prevents judge from creating spurious "Turn count requirement" criteria
- SDK goal achievement template rewritten to distinguish task methodology ("steps describing HOW to test") from success criteria ("conditions defining WHAT constitutes passing")

**Architecture simplification** (-313 lines, 2 deleted files):
- Inlined `GoalEvaluator` class (67-line wrapper around one method call) into agent.py
- Removed `ContextManager` class (3 "strategies" with identical implementations) — replaced with one line
- Deleted `system_assembly.py` passthrough module — renamed inner function to canonical name
- Demoted verbose executor logs from INFO to DEBUG

## Additional context

- Includes `turn_budget_validation.py` example with 14 test scenarios across 6 categories
- All 354 unit tests pass; all 14 validation scenarios pass
- Blog post in `playground/penelope/` elaborated with implementation war stories

## Testing

```bash
cd penelope && uv run pytest           # 354 tests
cd penelope && uv run python examples/turn_budget_validation.py --quick  # 6 core scenarios
cd penelope && uv run python examples/turn_budget_validation.py          # 14 full scenarios
```

### CODE REVIEW

File: penelope/src/rhesis/penelope/executor.py [hunk 1]
```diff
@@ -234,23 +193,22 @@ def execute_turn(
...
-                # We have history, use it
+                from rhesis.penelope.config import PenelopeConfig
+
                 prompt = user_prompt
-                # Use context manager for intelligent message selection
-                context_messages = ContextManager.select_context_messages(
-                    conversation_messages, strategy="recent"
-                )
-                for msg in context_messages:
+                max_messages = PenelopeConfig.DEFAULT_CONTEXT_WINDOW_MESSAGES
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply penelope/src/rhesis/penelope/executor.py:199]: Fixed in 6f1f788 — added `if max_messages > 0 else []` guard before slicing.

File: penelope/src/rhesis/penelope/agent.py [hunk 2]
```diff
@@ -564,6 +563,8 @@ def execute_test(
             restrictions=restrictions or "",
             context=str(context) if context else "",
             available_tools=available_tools_text,
+            min_turns=min_turns,
+            max_turns=max_turns or self.max_turns,
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply penelope/src/rhesis/penelope/agent.py]: Fixed in 6f1f788 — switched all three occurrences to `max_turns if max_turns is not None else self.max_turns`.

File: penelope/src/rhesis/penelope/utils.py [hunk 3]
```diff
@@ -178,10 +200,20 @@ def should_stop(self, state: TestState) -> tuple[bool, str]:
...
+        # This prevents giving up at min_turns when many turns remain.
         if isinstance(self.result.score, (int, float)) and self.result.score < 0.3:
-            reason = self.result.details.get("reason", "Low score after multiple attempts")
-            return True, f"Goal determined impossible: {reason}"
+            impossible_floor = self._get_impossible_floor()
+            if current_turns >= impossible_floor:
+                reason = self.result.details.get("reason", "Low score after multiple attempts")
+                return True, f"Goal determined impossible: {reason}"
+            else:
+                logger.info(
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply penelope/src/rhesis/penelope/utils.py]: Fixed in 6f1f788 — demoted to `logger.debug`.

File: sdk/src/rhesis/sdk/metrics/providers/native/templates/goal_achievement_prompt.jinja [hunk 4]
```diff
@@ -28,28 +28,30 @@ CONVERSATION GOAL (what should be achieved):
...
+   - NOTE: Numbered steps (e.g., "1. Ask about X, 2. Ask about Y") are task
+     steps, NOT turn count requirements. Do NOT create a turn count criterion
+     from numbered steps.
+
 3. **Partial Progress ≠ Success**: Even if some objectives were met:
-   - If the test stopped before completing required turns: FAILURE
    - If mandatory steps were skipped: FAILURE
    - Good partial results do NOT compensate for incomplete execution
 
 EVALUATION PRIORITY:
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply sdk/src/rhesis/sdk/metrics/providers/native/templates/goal_achievement_prompt.jinja:49]: The de-emphasis is intentional. The rule you're suggesting re-add already exists at line 40: `If actual turns < required turns: AUTOMATIC FAILURE (score = {{ min_score }})`. The change moved turn-count from priority #1 to #3 because the judge was over-indexing on it — treating numbered task steps ("1. Ask about X, 2. Ask about Y") as turn-count requirements and creating spurious criteria. The template now explicitly distinguishes task steps from turn-count directives (lines 41-43, 103-105, 114, 228).

File: penelope/src/rhesis/penelope/agent.py [hunk 5]
```diff
@@ -617,9 +618,22 @@ def execute_test(
             # Evaluate all SDK metrics
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply penelope/src/rhesis/penelope/agent.py:618]: There's no post-loop re-evaluation — the eval happens once per turn inside the loop, feeds `GoalAchievedCondition.update_result()` for stopping decisions, and appends to `state.metric_results` for score-over-time tracking. The per-turn cost is inherent to the design (we need a fresh evaluation after each new conversation turn to decide whether to stop). Caching would only help if there were a redundant final eval after the loop, which there isn't.
